feat(Story Reader Live2d)

kyokuheishin opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Implemented the Live2D version of Story Reader.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
N/A

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Since the Live2D reader has been unavailable for quite a long time, so I implemented it.

## How Has This Been Tested?
<!--- Please ensure your code is running good at least on following platform. -->
<!--- To test on mobile device, make sure they are connected to the same WiFi -->
<!--- with your developement machine. -->
<!--- If you can't test Safari browser, you can ignore them. -->

- [X] Chrome (Desktop)
- [X] Firefox (Desktop)


## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/16653295/153342933-f7c4807a-eead-4290-8790-f5a4a130f973.png)
![image](https://user-images.githubusercontent.com/16653295/153342947-98ec74f7-454c-4ea7-9750-fa651bd6790c.png)
![image](https://user-images.githubusercontent.com/16653295/153342959-3da67f20-f5d9-4565-a4ea-a847cfe2da8d.png)



## Details


### Features Implemented:
- Some Basic Special Effects (Changing background image, play music, black/white in/out, dialogue, etc.) 
- Live2D model displaying, changing motion/expression

### TODO:
- Better UI
- More Special Effects
- Lip sync

kyokuheishin committed:
feat(Story Reader Live2d): add StoryReaderLive2d.tsx

kyokuheishin committed:
feat(Story Reader Live2d): add useProcessedScenarioDataForLive2d in src/utils/index.ts

kyokuheishin committed:
feat(Story Reader Live2d): add StoryReaderLive2d.tsx

kyokuheishin committed:
feat(Story Reader Live2d): add StoryReaderContentLive2d.tsx

kyokuheishin committed:
feat(Story Reader Live2d): remove unnecessary parm "resourcesNeed"

Remove unnecessary parm "resourcesNeed" of function useProcessedScenarioDataForLive2d()

kyokuheishin committed:
feat(Story Reader Live2d): use useProcessedScenarioDataForLive2d()

kyokuheishin committed:
feat(Story Reader Live2d): use StoryReaderContentLive2d()

kyokuheishin committed:
feat(Story Reader Live2d): update function useProcessedScenarioDataForLive2d()

kyokuheishin committed:
feat(Story Reader Live2d): make live2d can be loaded

kyokuheishin committed:
feat(Story Reader Live2d): update router

kyokuheishin committed:
feat(Story Reader Live2d): update dependences

kyokuheishin committed:
feat(Story Player Live2d): let useProcessedScenarioDataForLive2d() also return resources need

kyokuheishin committed:
feat(Story Reader Live2d): update init and seek function of Story reader

kyokuheishin committed:
feat(Story Reader Live2d): update dependencies

kyokuheishin committed:
feat(Story Reader Live2d): fix typo

kyokuheishin committed:
feat(Story Reader Live2D): fix typo

kyokuheishin committed:
feat(Story Reader Live2d): make story player can be seeked by clicking the canvas

kyokuheishin committed:
feat(Story Reader Live2d): add dependencies

kyokuheishin committed:
feat(Story Reader Live2d): update function live2dScenarioPlayerSeek

kyokuheishin committed:
feat(Story Reader Live2d): add field seTypeId to case 'SnippetAction.SpecialEffect'

kyokuheishin committed:
feat(Story Reader Live2d): update dependencies

kyokuheishin committed:
feat(Story Reader Live2d): make useProcessedScenarioDataForLive2d return motionsNeed

kyokuheishin committed:
feat(Story Reader Live2d): use pixi-live2d-display to show live2d model

kyokuheishin committed:
feat(Story Reader Live2d): make character id can be gotten if there is no info in TalkCharacters

kyokuheishin committed:
feat(Story Reader Live2d): character motion, expression, telop, sekai in&out

kyokuheishin committed:
feat(Story Reader Live2d): make action object of FirstBackground has seTypeId

kyokuheishin committed:
feat(Story Reader Live2d): make router work

kyokuheishin committed:
feat(Story Reader Live2d): update parser

kyokuheishin committed:
feat(Story Reader Live2d): code cleanning

kyokuheishin committed:
feat(Story Reader Live2d): make useProcessedScenarioDataForLive2d() also returns FirstLayout

dnaroma commented:
Thanks for your efforts! I will review it asap

kyokuheishin committed:
feat(Story Reader Live2d): add more enum types of Layout

kyokuheishin committed:
feat(Story Reader Live2d): add case of LayoutType.Move

kyokuheishin committed:
feat(Story Reader Live2d): make LayoutType.Move work well

kyokuheishin committed:
feat(Story Reader Live2d): use live2dcubismcore in @sekai-world/find-live2d-v3/

dnaroma commented:
Hi, I reviewed your work, it's pretty awesome, but some suggestions:
1. Rename "Show" button to "Load".
2. Add progress bar to indicate loading progress.
3. Show "START!" on canvas before start instead of hint text above.
4. Some model keeps T-pose, motion not applied.
5. Add "Auto" button or checkbox to allow auto play.

Because too many bugs still exist, I can't approve it for merge it.

dnaroma reviewed (changes requested):
1. Rename "Show" button to "Load".
2. Add progress bar to indicate loading progress.
3. Show "START!" on canvas before start instead of hint text above.
4. Some model keeps T-pose, motion not applied.
5. Add "Auto" button or checkbox to allow auto play.

kyokuheishin committed:
feat(Story Reader Live2d): change label of button from "show" to "load"

kyokuheishin commented:
> 1. Rename "Show" button to "Load".
> 2. Add progress bar to indicate loading progress.
> 3. Show "START!" on canvas before start instead of hint text above.
> 4. Some model keeps T-pose, motion not applied.
> 5. Add "Auto" button or checkbox to allow auto play.

Sorry for not replying for so long, I've been a bit busy lately.
Thanks for the review! I've made several UI changes that you suggested.
But regarding the T-pose issue, I confirmed that the model loaded the action successfully and the issue only occurs when there are many ( n>= 8 ) characters appearing in a story at the same time, so I think it might be related to the third-party library I'm using. It may take me a while to verify my suspicions.

kyokuheishin committed:
fix(StoryReaderContentLive2d.tsx): change the behavior of dealing with live2d model

Create and destroy l2d model everytime it appears or disappears instead of loading them at the
beginning

kyokuheishin commented:
Still cannot find why loading 10 models could cause animation not updated, it may be a WebGL performance issue.

The solution now is, to create or destroy the model every time the scenario script tells it to appear or disappear, instead of keeping them in memory.

LazytownSpo666 commented:
Is Story Reader Live2D still working?

dnaroma commented:
> Is Story Reader Live2D still working?

Seems so

LazytownSpo666 commented:
> > Is Story Reader Live2D still working?
> 
> Seems so

if it does work, what will you (or anyone) need to run Story Reader Live2D?
(Also it's not working on any phone)

dnaroma commented:
> > > Is Story Reader Live2D still working?
> 
> > 
> 
> > Seems so
> 
> 
> 
> if it does work, what will you (or anyone) need to run Story Reader Live2D?
> 
> (Also it's not working on any phone)

Have you tested this branch? I'm not sure if it's still working. As long as you can use current Live2D viewer, you could use the Live2D story reader as well.

LazytownSpo666 commented:
Is this feature complete or not?

Also what happened to author of this pull request?!

AbnormalPoof commented:
I tried testing this a while back and only got a scuffed build working due to how old the commit is (Most stuff didn't work on my end). This might still work if it's adapted for the current version of Sekai Viewer, but I'm not really sure.

LazytownSpo666 commented:
> I tried testing this a while back and only got a scuffed build working due to how old the commit is (Most stuff didn't work on my end). This might still work if it's adapted for the current version of Sekai Viewer, but I'm not really sure.

Yep, you tried your best.
